[REEF-1392] Adding IObserver<ICloseEvent> for IMRU tasks

jwang98052 opened the pull request:
* IMRU tasks implement IObserver<ICloseEvent>
* Task will return after it receives the close event and completes the current iteration
* Update TaskStateMachine to allow transit from TaskWaitingForClose with CompletedTask event to TaskClosedByDriver state
* Adding test cases

JIRA: [REEF-1392](https://issues.apache.org/jira/browse/REEF-1392)
This closes

afchung commented:
I'll take a pass.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs:
```diff
@@ -200,13 +200,13 @@ internal void Reset()
         internal void RecordCompletedTask(ICompletedTask completedTask)
         {
-            RemoveRunningTask(completedTask.Id);
+            _runningTasks.Remove(completedTask.Id);
```
What's the difference?

jwang98052 replied:
RemoveRunningTask requires the task was in _runningTasks, meaning the task must be running before.
This method can be called in the scenario when driver sends event for the task to close, and the task returns from Call() gracefully, resulting driver receives ICompletedTask. In this case, the task was not in _runningTasks. Using _runningTasks.remove() fits the case even when the task is not in the collection.

dkm2110 replied:
@afchung , this is needed because there is a function below called CloseAllRunningTasks() which will send Close message to all the tasks and also removes taks from _runningTasks below. If @jwang98052  directly call RemoveRunningTask function she will get the exception..I am wondering if there is a cleaner way: Perhaps, do not remove from _runningTasks queue in CloseAllRunningTasks() and remove only once we receive failure or completed task messages. Problem  I faced while reviewing earlier also was that there were so many cases which makes me confused during reviewing. But I also do not have a cleaner solution :) .

markusweimer replied:
should `RemoveRunningTask` be removed, then?

jwang98052 replied:
RemoveRunningTask(string taskId) is used for the cases where Contains is required. As there is only one place left, I would remove it to avoid confusion.

markusweimer replied:
Maybe it needs a better name then? Right now, we have two ways to remove a taskid from the list of known running tasks, and they are confusingly similar.

markusweimer replied:
Or, alternatively, make `RemoveRunningTask` return a boolean which indicates whether the task was known.

jwang98052 replied:
I actually already removed RemoveRunningTask() method. 
At the place where contains is required, I do check it to ensure no mistake. 
At the places where contains is not required as the call may come from different states, I just call  _runningTasks.Remove(taskId); This method returns false it the element is not in the collection but no error.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -95,10 +98,35 @@ public byte[] Call(byte[] memento)
+        /// If it is closed event sent from driver, set _break to true so that to stop the iteration in Call method
+        /// </summary>
+        /// <param name="value"></param>
+        public void OnNext(ICloseEvent value)
```
rename `value` to avoid `value.Value.Value`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -95,10 +98,35 @@ public byte[] Call(byte[] memento)
+        public void OnNext(ICloseEvent value)
+        {
+            Logger.Log(Level.Info, "The task received close event with message:", value.Value.Value);
+            if (value.Value != null && value.Value.Value != null && Encoding.UTF8.GetString(value.Value.Value).Equals(TaskManager.CloseTaskByDriver))
```
use `value.IsPresent()` instead of `value.Value != null`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -104,19 +107,45 @@ public byte[] Call(byte[] memento)
+        /// Received close event. If it is closed command sent from driver, set _break to true to stop the iteration in Call method
+        /// </summary>
+        /// <param name="value"></param>
+        public void OnNext(ICloseEvent value)
```
Same as above.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -104,19 +107,45 @@ public byte[] Call(byte[] memento)
+        /// <param name="value"></param>
+        public void OnNext(ICloseEvent value)
+        {
+            if (value.Value != null && value.Value.Value != null && Encoding.UTF8.GetString(value.Value.Value).Equals(TaskManager.CloseTaskByDriver))
```
Same as above.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -104,19 +107,45 @@ public byte[] Call(byte[] memento)
+        {
+            if (value.Value != null && value.Value.Value != null && Encoding.UTF8.GetString(value.Value.Value).Equals(TaskManager.CloseTaskByDriver))
+            {
+                Logger.Log(Level.Info, "The task received close event with message:", value.Value.Value);
```
Missing format `{0}`. Also need to invoke with `string`, otherwise it will just print the address of the byte array.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -95,10 +98,35 @@ public byte[] Call(byte[] memento)
+        /// <param name="value"></param>
+        public void OnNext(ICloseEvent value)
+        {
+            Logger.Log(Level.Info, "The task received close event with message:", value.Value.Value);
```
Missing format {0}. Also need to invoke with string, otherwise it will just print the address of the byte array.
 Add a line note

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -33,14 +35,15 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         private readonly IReduceSender<TMapOutput> _dataReducer;
         private readonly IMapFunction<TMapInput, TMapOutput> _mapTask;
         private readonly bool _invokeGC;
+        private bool _break;
```
`volatile`

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -34,14 +36,15 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         private readonly IBroadcastSender<MapInputWithControlMessage<TMapInput>> _dataAndControlMessageSender;
         private readonly IUpdateFunction<TMapInput, TMapOutput, TResult> _updateTask;
         private readonly bool _invokeGC;
+        private bool _break;
```
`volatile`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +347,52 @@ public void OnError(Exception error)
 
+        private sealed class BreakTestTask : ITask, IObserver<ICloseEvent>
+        {
+            private bool _break;
```
`volatile`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -69,6 +69,22 @@ public void TestStopTaskOnLocalRuntime()
+            ValidateMessageSuccessfullyLoggedForDriver(CompletedValidationMessage, testFolder, 1);
+            var messages = new List<string>();
+            messages.Add(DisposeMessageFromDriver);
+            ValidateMessageSuccessfullyLogged(messages, "Node-*", EvaluatorStdout, testFolder, 2);
```
This only validates that the dispose message is received. This doesn't validate that the `Task` actually breaks.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +347,52 @@ public void OnError(Exception error)
+
+            public byte[] Call(byte[] memento)
+            {
+                while (!_break && _iterate < 100)
```
What is `_iterate` for? Why isn't it a local variable??

jwang98052 replied:
In case the task doesn't receive the close event, we don't want the task runs forever. The iterate can be local in this test.

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +347,52 @@ public void OnError(Exception error)
+                {
+                    Logger.Log(Level.Info, "Closed event received in task:" + Encoding.UTF8.GetString(value.Value.Value));
+                    _break = true;
+                }
```
Are we expecting a different kind of message? If not throw.

jwang98052 replied:
done

afchung commented:
Done with a first pass.

jwang98052 commented:
@afchung I have addressed your review comments.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +100,46 @@ public byte[] Call(byte[] memento)
+
+            if (!_stopped)
+            {
+                throw new SystemException(TaskManager.TaskKilledByDriver);
```
`IMRUSystemException`?

jwang98052 replied:
IMRUSystemException is used for unknown exceptions in IMRU system. For exceptions in fault tolerant, I would add IMRUTaskSystemException.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -33,24 +36,29 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         /// </summary>
         /// <param name="mapTask">The MapTask hosted in this REEF Task.</param>
         /// <param name="groupCommunicationsClient">Used to setup the communications.</param>
+        /// <param name="enforceCloseTimeout">Timeout to enforce the task to close if receing task close event</param>
```
receiving

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -34,27 +37,32 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
             IUpdateFunction<TMapInput, TMapOutput, TResult> updateTask,
             IGroupCommClient groupCommunicationsClient,
             IIMRUResultHandler<TResult> resultHandler,
+            [Parameter(typeof(EnforceCloseTimeout))] int enforceCloseTimeout,
```
`long`.

afchung replied:
Is this in Milliseconds?

jwang98052 replied:
yes. changed.

afchung replied:
Please change `EnforceCloseTimeout` to `EnforceCloseTimeoutMilliseconds`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -104,19 +113,53 @@ public byte[] Call(byte[] memento)
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(_enforceCloseTimeout);
```
Use `TimeSpan.FromMilliseconds`.

afchung replied:
I don't think this is a good idea. Using real-time constraints may surface different behaviors on different machines.

markusweimer replied:
+1

jwang98052 replied:
Here is the scenario: driver sends close event to task for it to e closed, however, the task is hung in Call() method for a blocking call. There is no way to signal the close handler to take some action. In this case, we would just want to enforce the task to close by throwing an exception. In this case, the accuracy of the time is not very critical as we just want to eventually close the task after waiting for some time if task is not returned from Call() yet. 

Please describe what's the potential problem here and any better idea.

jwang98052 replied:
TimeSpan.FromMilliseconds requires double. What is the advantage of using it? More accurate?

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/EnforceCloseTimeout.cs:
```diff
@@ -0,0 +1,30 @@
+    /// After specified time out, if the task is still not closed, the close handler will throw exception enforcing the task to close. 
+    /// </summary>
+    [NamedParameter("Enfore the task to close time out in millisecond", "EnforceCloseTimeout", "1000")]
+    internal sealed class EnforceCloseTimeout : Name<int>
```
Should this really be `internal`?

jwang98052 replied:
IMRU task configuration is set in IMRUDriver that is in IMRU project. If we would like client to set it, IMRU client is also in IMRU project.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/EnforceCloseTimeout.cs:
```diff
@@ -0,0 +1,30 @@
+namespace Org.Apache.REEF.IMRU.OnREEF.Parameters
+{
+    /// <summary>
+    /// When driver sends clsoe event to a task, it would expect the task to clsoe gracefully. 
```
close, for both instances.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +386,85 @@ public void OnError(Exception error)
+                    if (_enforceClose)
+                    {
+                        Thread.Sleep(_enforceCloseTimeout);
+                        if (!_stoped)
```
`_stopped`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +386,85 @@ public void OnError(Exception error)
+                    _break = true;
+                    if (_enforceClose)
+                    {
+                        Thread.Sleep(_enforceCloseTimeout);
```
What is the purpose of this timeout? Do we have to have it?

jwang98052 replied:
Yes.

The idea case is to inform Call() to finish the current iteration then return. However, in group communication case, when one of the task fails, all the other tasks are hung in a blocking call. We need a back up plan to enforce the closing. 

This issue has been discussed in Jira.

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +386,85 @@ public void OnError(Exception error)
+            {
+                int iterate = 1;
+
+                while (!_break && iterate < 100)
```
What is `iterate` for again?

afchung replied:
Our tests should be deterministic and should never rely upon iteration count and timeouts.

jwang98052 replied:
This is to simulate the real scenario. In the real scenario, we want to complete current iteration then break.

dkm2110 replied:
I am with @afchung  on this. Timeouts in these tests make me nervous.

jwang98052 replied:
I have modified the test code to use CountdownEvent to make it deterministic.

afchung commented:
@jwang98052 done with another pass.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +100,46 @@ public byte[] Call(byte[] memento)
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(_enforceCloseTimeout);
```
What is this used for? Real time units are always dangerous.

jwang98052 replied:
As we discussed in Jira, this is for enforcing to close after waiting for a configured time.

markusweimer commented:
The JIRA discussion on this hasn't converged yet. Let's resolve what to do about the state management over there.

jwang98052 commented:
I have created REEF-1404 for task state maintenance and preservation for IMRU Task at Evaluator side.

jwang98052 commented:
@afchung @markusweimer I have addressed the review comments and left some questions to clarify.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -64,6 +72,7 @@ internal sealed class UpdateTaskHost<TMapInput, TMapOutput, TResult> : ITask
             _dataReceiver = cg.GetReduceReceiver<TMapOutput>(IMRUConstants.ReduceOperatorName);
             _invokeGC = invokeGC;
             _resultHandler = resultHandler;
+            _enforceCloseTimeout = enforceCloseTimeout;
```
`_enforceCloseTimeoutMilliseconds`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/EnforceCloseTimeout.cs:
```diff
@@ -0,0 +1,30 @@
+    /// After specified time out, if the task is still not closed, the close handler will throw exception enforcing the task to close. 
+    /// </summary>
+    [NamedParameter("Enfore the task to close time out in millisecond", "EnforceCloseTimeout", "1000")]
+    internal sealed class EnforceCloseTimeout : Name<int>
```
`EnforceCloseTimeoutMilliseconds`.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/EnforceCloseTimeout.cs:
```diff
@@ -0,0 +1,30 @@
+    /// When driver sends close event to a task, it would expect the task to close gracefully. 
+    /// After specified time out, if the task is still not closed, the close handler will throw exception enforcing the task to close. 
+    /// </summary>
+    [NamedParameter("Enfore the task to close time out in millisecond", "EnforceCloseTimeout", "1000")]
```
Is 1 second long enough?

jwang98052 replied:
change to 5 :)

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -69,6 +74,40 @@ public void TestStopTaskOnLocalRuntime()
+        /// This test is to close a running task and enforce it to break and return after the current iteration
+        /// </summary>
+        [Fact]
+        public void TestBreakTaskOnLocalRuntime()
```
Can you use a real IMRU task? Writing an emulation Task requires you to maintain both.

afchung replied:
It also obscures what you are testing for.

jwang98052 replied:
It is really hard as IMRU tasks require a lot injectable classes in constructors and each has nested injectable classes. It is also very complex to mock as those will be used in Call method. Too much to mock.

dkm2110 replied:
Can't you take existing IMRU Tests and create their modified version for this?

afchung replied:
+1

jwang98052 replied:
REEF-1408 is to test real IMRU task. It is pending for sending PR as it has dependency on this one. As it is not small test case and involves several classes, I would prefer to make it as a separate PR.

afchung commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +388,85 @@ public void OnError(Exception error)
+                        Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeout));
+                        if (!_stopped)
+                        {
+                            Logger.Log(Level.Info, EnforceToCloseMessage);
```
This is not guaranteed to be logged.

jwang98052 replied:
Ok. I replaced sleep in the call method with CountdownEvent.

afchung commented:
@jwang98052 Done with another round of review.

jwang98052 commented:
@afchung Thanks for the review! I have addressed your review comments. Please look at it again. Thanks!

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Parameters/EnforceCloseTimeoutMilliseconds.cs:
```diff
@@ -0,0 +1,30 @@
+    /// When driver sends close event to a task, it would expect the task to close gracefully. 
+    /// After specified time out, if the task is still not closed, the close handler will throw exception, enforcing the task to close. 
+    /// </summary>
+    [NamedParameter("Enfore the task to close time out in millisecond", "EnforceCloseTimeout", "5000")]
```
Change to "Enforce the task to close after waiting for this much time (in milliseconds).

jwang98052 replied:
done

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -33,24 +36,29 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         private readonly IReduceSender<TMapOutput> _dataReducer;
         private readonly IMapFunction<TMapInput, TMapOutput> _mapTask;
         private readonly bool _invokeGC;
+        private volatile bool _break;
```
nit: is explicit initialization preferrable?

jwang98052 replied:
Yes, make the code readable.

markusweimer replied:
Also, `break` is a bit generic a name.

jwang98052 replied:
changed.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +100,46 @@ public byte[] Call(byte[] memento)
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```
Why is this needed at all? Why does task need this time?

jwang98052 replied:
After setting the -break to true, we would like Call() to complete the current iteration then return. We should give some time for the the task to gracefully return by itself. But if for whatever reason, blocking call can be one of the reasons, the task is hung and it cannot reach to the next iteration and cannot break by itself, after waiting for specified time, we have to enforce the task to stop. That is the reason we need to sleep here.

dkm2110 replied:
So is it fair to say this is a hack we are putting since GC currently blocks if underlying stream closes. @markusweimer  any comments on this?

jwang98052 replied:
I won't say it is a hack. When driver ask a task to close, for whatever reason, if the task cannot close in the specified time frame, enforce close should happen. This is to make the code is robust. Even after we resolve that blocking call issue, we still need some back up plan to ensure sure the task is closed in case it cannot do it by itself.

dkm2110 replied:
there are only three reasons it will not come out:
a) GC does not throw exceptions when stream fails - this is limitation of current REEF and need to be addressed.
b) Map or Update function hangs inside - This is a bug in user code and we are not responsible for it and should not care for it.
c) Any other reason - This is either a bug in REEF or our Host tasks and is again a REEF problem.

If anything, these sleeps will hide any real bugs in future.

markusweimer replied:
This wait is problematic. We even wait when the task exits instantaneously. This should be rewritten with a while / sleep loop or the better version of it .NET has to offer.

jwang98052 replied:
Yeak, we can reduce sleep time but add retry to make it to finish quicker. Done.

jwang98052 replied:
I created REEF-1407 to track the work for task blocking call. We will revisit here once that work is done.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -104,19 +113,53 @@ public byte[] Call(byte[] memento)
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
+            }
+            Thread.Sleep(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
```
Again same comment as above. My understanding is that you need this sleep since after breaking from Task call, it needs to dispose result handler and also set _stopped right? If you need task to do somethings and want CloseEvent to wait, why not use Semaphores or related structures that are much cleaner?

jwang98052 replied:
After close handler set _break to true, it would expect the task to return by itself. The code flow is not always back to the close handler. 
Close handler wait and then enforce to close is a backup plan, it is for the case if the task hung and cannot return from Call() by itself. In this case, we must enforce to close. 
I am not sure how Semaphores can be used in this scenario. Can you elaborate?

dkm2110 replied:
They cannot be. I was wrong there. Real problem is that GC does not throw exception. Hence, you cannot come out of loop in case network issues are there and hence this hack is needed. Real solution is to make sure GC throws appropriate exception.

jwang98052 replied:
That's right. 

But still we need a backup plan because we may miss some other scenarios. Because in any case, the task must be stopped in order to submit a new one, no matter how it is stopped.

dkm2110 replied:
Please see my comment above regarding this.

dkm2110 commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -322,6 +388,92 @@ public void OnError(Exception error)
             }
         }
 
+        private sealed class BreakTestTask : ITask, IObserver<ICloseEvent>
```
Can you add documentations to all these tasks. Its very difficult to see clearly what each one is trying to do. Also can you expand on the names of tasks to make them more informative?

jwang98052 replied:
I updated comments and task name for this task. For other task, I added comments. Rename would cause more changes everywhere that would disturb people from reviewing this PR.

dkm2110 replied:
I do not see why this should be problem for other reviewers if names are better..Anycase.

dkm2110 commented:
@jwang98052 I have left some comments. Main concerns are timeouts introduced in *Host tasks and tests since they actually do not check the *Host tasks. Can't we take the current IMRU BroadcastAndReudce or MapperCount test and modify them for this purpose or even something simpler?

jwang98052 commented:
@dkm2110 Regression testing for IMRU functional tests passed. 
To reach the code path in testing, we need to send close event to the host tasks, that requires IMRU driver change, but that is REEF source code and we are not suppose to change it for test purpose. Can we only change the test code with IMRU driver and IMRUClient unchanged to achieve the test purpose,? Any suggestion?

dkm2110 commented:
@jwang98052  Why do you need IMRU driver to check UpdateTaskHost and MapTaskHost? They are used in IMRU but do not need IMRU driver for testing. You can take Group Communication functional tests and there as master and slave tasks inject these *Host tasks. You simply need to define Map and Update functions needed by *Host tasks. Map function can be Identity and Update function can simply transmit an integer. Basically Master and Slave tasks in GC become Update and Map functions. Then modify the driver so that when you have all the RunningTasks, you send close message to one of the tasks. In one test send it to MapTaskHost and in other to UpdateTaskHost.

Most of the code already exists and just needs to be massaged a little.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/IMRUTaskSystemException.cs:
```diff
@@ -0,0 +1,42 @@
+        /// <summary>
+        /// Constructor. A serialiable exception object that represents a task system error.
+        /// </summary>
+        public IMRUTaskSystemException(string message)
```
`internal`? Also, what is a `task system error`? That isn't clear from the docs.

jwang98052 replied:
class internal override method.

jwang98052 replied:
Added doc.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/IMRUTaskSystemException.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+        }
+
+        public IMRUTaskSystemException(SerializationInfo info, StreamingContext context)
```
`internal`?

jwang98052 replied:
Call internal override method. Also this method is required by [Serializable], we need to follow its signature. .

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -33,24 +36,29 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
     /// <typeparam name="TMapInput">Map input</typeparam>
     /// <typeparam name="TMapOutput">Map output</typeparam>
-    internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+    internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask, IObserver<ICloseEvent>
```
This class now has a mutable member. Please document its threading properties with the attributes we introduced.

jwang98052 replied:
you mean [NotThreadSafe]?

jwang98052 replied:
I changed to use Interlocked for the variables, then this attribute is not needed, right?

markusweimer replied:
It is, but now we believe the class to be `[ThreadSafe]`, right?

jwang98052 replied:
Done.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +100,46 @@ public byte[] Call(byte[] memento)
+            if (closeEvent.Value.IsPresent() && Encoding.UTF8.GetString(closeEvent.Value.Value).Equals(TaskManager.CloseTaskByDriver))
+            {
+                Logger.Log(Level.Info, "The task received close event with message: {0}.", Encoding.UTF8.GetString(closeEvent.Value.Value));
+                _break = true;
```
This is changed on a different thread than read. Is this safe?

jwang98052 replied:
I can use Interlocked.Exchange if we want to lock it.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs:
```diff
@@ -34,27 +37,32 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         private readonly IBroadcastSender<MapInputWithControlMessage<TMapInput>> _dataAndControlMessageSender;
         private readonly IUpdateFunction<TMapInput, TMapOutput, TResult> _updateTask;
         private readonly bool _invokeGC;
+        private volatile bool _break;
```
Please document the threading behavior and choose more descriptive names.

jwang98052 replied:
done

markusweimer commented:
I have done a pass.

jwang98052 commented:
@markusweimer @dkm2110 I have addressed your review comments. Please look at it again. Thanks! 

I have spent almost a day to create a test driver for IMRU tasks. It is not simply to replace task class name in task configuration and add map.update function config with existing BroadCastReduce test driver, it involves too many configurations including group communication config, dataset/IInputPartition config, all the codec config, group communication client config, I was almost rewrite another IMRU driver but still end up wrong configurations. I feel we might just use IMRU driver and modify it for testing, but I am not sure if we should duplicate the prod code in testing, especially we are rewriting it now. For now, I don't think we should spend so much time to rewrite some thing for this type of testing. It should be covered by E2E testing.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -33,24 +36,32 @@ namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
         private readonly IReduceSender<TMapOutput> _dataReducer;
         private readonly IMapFunction<TMapInput, TMapOutput> _mapTask;
         private readonly bool _invokeGC;
+        private long _toClose = 0;
```
Should this be `_taskToClose` and `_tasksStopped`?

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -68,7 +81,7 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
         public byte[] Call(byte[] memento)
         {
-            while (true)
+            while (Interlocked.Read(ref _toClose) == 0)
```
Is this safe? What if two threads end up here at around the same time?

jwang98052 replied:
My understanding is Interlocked only lock that variable when accessing it. Once entering the while loop, the lock should have been released.

markusweimer commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +104,52 @@ public byte[] Call(byte[] memento)
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            if (closeEvent.Value.IsPresent() && Encoding.UTF8.GetString(closeEvent.Value.Value).Equals(TaskManager.CloseTaskByDriver))
```
`Encoding.UTF8.GetString(closeEvent.Value.Value)` is computed twice here. Maybe use a local variable for it?

jwang98052 replied:
Sure.

markusweimer commented:
I've done a pass.

jwang98052 commented:
@markusweimer I have addressed your review comments. Please look at it again. 
For IMRUTask testing, I have come out a solution and tests cases. As it involves several classes, I would like to have a separate PR for it. REEF-1408 is created to track it.

jwang98052 commented:
@afchung @dkm2110 @markusweimer Do you still have any comments for this PR?

dkm2110 commented:
@jwang98052 except test comments LGTM. @markusweimer @afchung please do not merge this one right away. @jwang98052  once everybody sign off please leave a message and I will test with 1000 evaluators on cluster since the changes here include the ones in current IMRUDriver too.

jwang98052 commented:
@dkm2110 it would be nice if you can do a regression testing. Thanks!

I already have REEF-1408 ready in my private branch to test IMRUTasks and all test cases have passed. I didn't send it for review because it has dependency on this PR. But if you would like to review it now, I can also send it for CR. It cannot be merged until this PR is merged.

dkm2110 commented:
@jwang98052  I will do it once everybody signs off. I do not want to do it again and again.

afchung commented:
I'll do another pass.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
         private readonly bool _invokeGC;
 
         /// <summary>
+        /// When receiving close event, this varable will be set to 1. At the begining of the task iteration, 
```
beginning

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
 
         /// <summary>
+        /// When receiving close event, this varable will be set to 1. At the begining of the task iteration, 
+        /// if this varibale is set to 1, it will break from the loop and return from the Call() method. 
```
variable

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+        /// When receiving close event, this varable will be set to 1. At the begining of the task iteration, 
+        /// if this varibale is set to 1, it will break from the loop and return from the Call() method. 
+        /// </summary>
+        private long _taskToClose = 0;
```
`_shouldCloseTask`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+        /// If after retry and sleep, if this variable is still 0, the task could be hung, then tehe close handler will throw 
+        /// IMRUTaskSystemException to enforce the task to be closed. 
+        /// </summary>
+        private long _taskStopped = 0;
```
`_isTaskStopped`

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+        private long _taskToClose = 0;
+
+        /// <summary>
+        /// Before the task is returned, this varaible is set to 1. If the return is ordered by the close handler, 
```
variable

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+
+        /// <summary>
+        /// Before the task is returned, this varaible is set to 1. If the return is ordered by the close handler, 
+        /// the close handler will keep checking this variable. If it is 1, meaning the yask has been returned byitself, then do nothing. 
```
task has been returned by itself

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -43,14 +48,42 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask
+        /// <summary>
+        /// Before the task is returned, this varaible is set to 1. If the return is ordered by the close handler, 
+        /// the close handler will keep checking this variable. If it is 1, meaning the yask has been returned byitself, then do nothing. 
+        /// If after retry and sleep, if this variable is still 0, the task could be hung, then tehe close handler will throw 
```
the

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs:
```diff
@@ -91,14 +126,53 @@ public byte[] Call(byte[] memento)
+            }
+
+            var retry = 0;
+            while (Interlocked.Read(ref _taskStopped) == 0 && retry < _enforceCloseMaxRetryCount)
```
This seems like something you should use a `WaitHandle` for.

afchung commented:
Looks good overall to me as well, but I also agree with @dkm2110 on the test comment. I think we should be using a real IMRU Task or find a way to mock it out instead of using an ad-hoc testing class.

jwang98052 commented:
@afchung I have addressed your comments. Please look at it again. Thanks! 

The tests for IMRU tasks is addressed in REEF-1408, pending for REEF-1403 deadlock solution to check in.  The tests have been passed. I can send for review first.

jwang98052 commented:
@afchung I have addressed your comments. Could you please have another look? Thanks!

afchung commented:
@jwang98052 your change was not pushed.

jwang98052 commented:
@afchung Thanks for the reminder. I thought I have pushed in Saturday.

afchung commented:
Mostly looks good to me, but I'm still a bit iffy about not using a real IMRU Task.

jwang98052 commented:
We have IMRU regression tests that use IMRUTasks and the tests passed. 
I do have REEF-1408 for more IMRU testing including the one for IMRU task close event. It is just a sequence of check in to avoid too big PR, PR dependency and endless review. 

@markusweimer please do another review if you would like to look at the fix for the review comments.

markusweimer commented:
I agree with @afchung and @dkm2110 on the testing. We need to have good testing for this functionality. However, building a test framework for IMRU will probably make this change much bigger. Hence, I think it might be better to do as a separate code review as part of REEF-1408.

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/Bridge/TestCloseTask.cs:
```diff
@@ -69,6 +74,40 @@ public void TestStopTaskOnLocalRuntime()
+            messages.Add(DisposeMessageFromDriver);
+            messages.Add(EnforceToCloseMessage);
+            ValidateMessageSuccessfullyLogged(messages, "Node-*", EvaluatorStdout, testFolder, -1);
+            ////CleanUp(testFolder);
```
Why the commented out code?

jwang98052 replied:
oh, forgot to recover after testing. Good catch.

markusweimer commented:
I've done another pass leaving one comment.

dkm2110 commented:
@markusweimer @afchung  In that case I believe there are no further comments. Once you guys sign off, I will test on cluster and merge.

dkm2110 commented:
Also I assume REEF-1408 would be done immediately after this to maintain continuity?

jwang98052 commented:
@afchung already puts his final comment, @markusweimer final one is for the "////" which I have removed. Do you want them to say something like sign off? 

As soon as this one is merged to master, I will send REEF-1408 for review. It has dependency on this PR.

jwang98052 commented:
@dkm2110 in fact, since current IMRU driver doesn't send task close event anywhere, the change shouldn't impact any test result. But no harm to give a try.

dkm2110 commented:
@jwang98052  I will test and merge today.

jwang98052 commented:
@dkm2110 Thanks!

dkm2110 commented:
@jwang98052 Can you do merge and rebase with latest REEF bits. When I tried to pull the branch I got conflicts. Let me know when it is done.

jwang98052 commented:
@dkm2110 I have rebased it and tested in my local.

jwang98052 commented:
AppVeyor testing passed in my repository. 
https://ci.appveyor.com/project/jwang98052/reef/build/7-REEF-1392
